src/Surfnet/StepupGateway/GatewayBundle/Tests/Entity/SecondFactorTest.php

```diff
@@ -0,0 +1,159 @@
+            new Loa($requiredLoa, 'loa_identifier'),
+            $vettingTypeService
+        );
+    }
+
+    /**
+     * @dataProvider provideLoaLevelExpectations
+     */
+    public function test_get_loa(string $tokenType, bool $isIdentityVetted, float $expectedLoa)
+    {
+        $token = $this->buildSecondFactor($tokenType, $isIdentityVetted);
+        $service = $this->buildSecondFactorTypeService();
+        $this->assertEquals($expectedLoa, $token->getLoaLevel($service));
+    }
+
+    public function provideIdentityVettedTestData()
+    {
+        return [
+            'identity vetted yubikey satisfies loa3' => ['yubikey', 3.0, true],
+            'identity vetted demo_gssp satisfies loa3.0' => ['demo_gssp', 3.0, true],
+            'identity vetted tiqr does not satisfy loa3.0' => ['tiqr', 3.0, false],
+            'identity vetted sms does not satisfy loa3.0' => ['sms', 3.0, false],
+
+            'identity vetted yubikey satisfies loa2.0' => ['yubikey', 2.0, true],
+            'identity vetted demo_gssp satisfies loa2.0' => ['demo_gssp', 2.0, true],
+            'identity vetted tiqr satisfies loa2.0' => ['tiqr', 2.0, true],
+            'identity vetted sms satisfies loa2.0' => ['sms', 2.0, true],
+
+            'identity vetted yubikey satisfies loa1.5' => ['yubikey', 1.5, true],
+            'identity vetted demo_gssp satisfies loa1.5' => ['demo_gssp', 1.5, true],
+            'identity vetted tiqr satisfies loa1.5' => ['tiqr', 1.5, true],
+            'identity vetted sms satisfies loa1.5' => ['sms', 1.5, true],
+
+            'self-asserted yubikey does not satisfy loa3.0' => ['yubikey', 3.0, false, false],
+            'self-asserted demo_gssp does not satisfy loa3.0' => ['demo_gssp', 3.0, false, false],
+            'self-asserted tiqr does not satisfy loa3.0' => ['tiqr', 3.0, false, false],
+            'self-asserted sms does not satisfy loa3.0' => ['sms', 3.0, false, false],
+
+            'self-asserted yubikey does not satisfy loa2.0' => ['yubikey', 2.0, false, false],
+            'self-asserted demo_gssp does not satisfy loa2.0' => ['demo_gssp', 2.0, false, false],
+            'self-asserted tiqr does not satisfy loa2.0' => ['tiqr', 2.0, false, false],
+            'self-asserted sms does not satisfy loa2.0' => ['sms', 2.0, false, false],
+
+            'self-asserted yubikey satisfies loa1.5' => ['yubikey', 1.5, true, false],
+            'self-asserted demo_gssp satisfies loa1.5' => ['demo_gssp', 1.5, true, false],
+            'self-asserted tiqr satisfies loa1.5' => ['tiqr', 1.5, true, false],
+            'self-asserted sms satisfies loa1.5' => ['sms', 1.5, true, false],
+        ];
+    }
+
+    public function provideErroneousIdentityVettedTestData()
+    {
+        return [
+            'unknown token type' => ['tjoepiekey', 3.0, "The Loa level of this type: tjoepiekey can't be retrieved."],
+            'unknown loa level' => ['yubikey', 0, 'Unknown loa level "0", known levels: "1", "1.5", "2", "3"'],
```

Comment:
An idea to add some more edge cases like 1.4, 1.6, 4?

Reply:
All I do is deliver edge-cases to you guys ;) I'm all in favor

Reply:
Haha, I attempted to test the possible exception avenues. But I can spice them up a notch :)

Reply:
I added some tests, and lo and behold. I found a bug in the Stepup-Bundle. The exception message casts the loA level to an integer value. Resulting in a message like (when specifying a level 3.5 token): 

'Invalid LoA level 3, valid levels are 1, 2, 3, 1.5'

Given this is a bug in another component, I'm not fixing this here. But instead add a report of this issue on the stepup backlog.